test: flaky queue/Example_connectionPool on macOS

https://github.com/tarantool/go-tarantool/actions/runs/4511463511/jobs/7944401745
```
2023-03-24T13:52:48.0576220Z === RUN   Example_connectionPool
2023-03-24T13:52:49.1545340Z 2023/03/24 13:52:49 tarantool: storing connection to 127.0.0.1:3015 canceled: .../.rocks/share/tarantool/queue/abstract/queue_session.lua:223: The UUID f16c5fac-9d33-4640-8494-1043d26b4b80 is unknown. (0x20)
2023-03-24T13:52:49.1556050Z --- FAIL: Example_connectionPool (1.10s)
```

https://github.com/tarantool/go-tarantool/actions/runs/4512321687/jobs/7948237951
```
--- FAIL: Example_connectionPool (0.01s)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x48 pc=0x438ed37]

goroutine 1 [running]:
testing.(*InternalExample).processRunResult(0xc0001e95a8, 0xc0000289f0, 0x28, 0xcbff1f, 0x4438b40, 0x4825660, 0x4015a70)
	/Users/runner/hostedtoolcache/go/1.13.15/x64/src/testing/example.go:89 +0x63f
testing.runExample.func2(0xc0ff93057598e2e0, 0x6ec8ac3cc, 0x4831f20, 0xc0001184d0, 0xc0000aa008, 0xc000131440, 0xc0001e95a8, 0xc0001e95d8)
	/Users/runner/hostedtoolcache/go/1.13.15/x64/src/testing/run_example.go:58 +0x105
panic(0x4438b40, 0x4825660)
	/Users/runner/hostedtoolcache/go/1.13.15/x64/src/runtime/panic.go:679 +0x1b2
github.com/tarantool/go-tarantool.(*Connection).loadSchema(0xc000296000, 0x44d11b8, 0xc000296000)
	/Users/runner/work/go-tarantool/go-tarantool/tarantool/go-tarantool/schema.go:328 +0x397
github.com/tarantool/go-tarantool.Connect(0x44b6cd8, 0xe, 0x0, 0x0, 0x0, 0x3b9aca00, 0x0, 0x0, 0x44b3ee0, 0x4, ...)
	/Users/runner/work/go-tarantool/go-tarantool/tarantool/go-tarantool/connection.go:430 +0x65d
github.com/tarantool/go-tarantool/connection_pool.(*ConnectionPool).fillPools(0xc000098b60, 0xc00010ebd0, 0x44b6cd8, 0xe, 0xc00010ec89)
	/Users/runner/work/go-tarantool/go-tarantool/tarantool/go-tarantool/connection_pool/connection_pool.go:905 +0x1fd
github.com/tarantool/go-tarantool/connection_pool.ConnectWithOpts(0xc00010f2d0, 0x2, 0x2, 0x0, 0x0, 0x0, 0x3b9aca00, 0x0, 0x0, 0x44b3ee0, ...)
	/Users/runner/work/go-tarantool/go-tarantool/tarantool/go-tarantool/connection_pool/connection_pool.go:146 +0x5fb
github.com/tarantool/go-tarantool/queue_test.Example_connectionPool()
	/Users/runner/work/go-tarantool/go-tarantool/tarantool/go-tarantool/queue/example_connection_pool_test.go:151 +0x2d9
testing.runExample(0x44ba4e8, 0x16, 0x44d1208, 0x44cb6ba, 0x9a, 0x0, 0x0)
	/Users/runner/hostedtoolcache/go/1.13.15/x64/src/testing/run_example.go:62 +0x209
testing.runExamples(0xc0001e9770, 0x482b820, 0x3, 0x3, 0x101)
	/Users/runner/hostedtoolcache/go/1.13.15/x64/src/testing/example.go:44 +0x1a8
testing.(*M).Run(0xc00012a200, 0x0)
	/Users/runner/hostedtoolcache/go/1.13.15/x64/src/testing/testing.go:1118 +0x203
github.com/tarantool/go-tarantool/queue_test.runTestMain(0xc00012a200, 0x0)
	/Users/runner/work/go-tarantool/go-tarantool/tarantool/go-tarantool/queue/queue_test.go:952 +0x530
github.com/tarantool/go-tarantool/queue_test.TestMain(0xc00012a200)
	/Users/runner/work/go-tarantool/go-tarantool/tarantool/go-tarantool/queue/queue_test.go:956 +0x2b
main.main()
	_testmain.go:84 +0x135
FAIL	github.com/tarantool/go-tarantool/queue	30.049s
```

```
=== RUN   Example_connectionPool
--- FAIL: Example_connectionPool (0.03s)
got:
Master 127.0.0.1:3014 is ready to work!
A Queue object is ready to work.
Send data: test_data
Unable to set cluster roles: Space '_queue_taken_2_mgr' already exists (0xa)
want:
Master 127.0.0.1:3014 is ready to work!
A Queue object is ready to work.
Send data: test_data
Master 127.0.0.1:3015 is ready to work!
Got data: test_data
```

```
=== RUN   Example_connectionPool
--- FAIL: Example_connectionPool (5.01s)
got:
Master 127.0.0.1:3014 is ready to work!
A Queue object is ready to work.
Send data: test_data
Master 127.0.0.1:3015 is ready to work!
Unable to got task: can't find rw instance in pool
want:
Master 127.0.0.1:3014 is ready to work!
A Queue object is ready to work.
Send data: test_data
Master 127.0.0.1:3015 is ready to work!
Got data: test_data
```

```
=== RUN   Example_connectionPool
2023/05/12 07:50:00 tarantool: storing connection to 127.0.0.1:3015 canceled: .../.rocks/share/tarantool/queue/abstract/queue_session.lua:223: The UUID eff1bc1e-6e56-4c1b-bc6e-9c635a5fb4fd is unknown. (0x20)
--- FAIL: Example_connectionPool (5.08s)
got:
Master 127.0.0.1:3014 is ready to work!
A Queue object is ready to work.
Send data: test_data
Unable to re-identify in the pool: .../.rocks/share/tarantool/queue/abstract/queue_session.lua:223: The UUID eff1bc1e-6e56-4c1b-bc6e-9c635a5fb4fd is unknown. (0x20)
want:
Master 127.0.0.1:3014 is ready to work!
A Queue object is ready to work.
Send data: test_data
Master 127.0.0.1:3015 is ready to work!
Got data: test_data
FAIL
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -18,6 +18,8 @@ Versioning](http://semver.org/spec/v2.0.0.html) except to the first release.
 
 ### Changed
 
+- queue module version bumped to 1.3.0 (#278)
+
 ### Fixed
 
 - Several non-critical data race issues (#218)

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -24,6 +24,7 @@ Versioning](http://semver.org/spec/v2.0.0.html) except to the first release.
 - ConnectionPool does not properly handle disconnection with Opts.Reconnect
   set (#272)
 - Watcher events loss with a small per-request timeout (#284)
+- Connect() panics on concurrent schema update (#278)
 
 ## [1.10.0] - 2022-12-31
 

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -27,6 +27,7 @@ Versioning](http://semver.org/spec/v2.0.0.html) except to the first release.
   set (#272)
 - Watcher events loss with a small per-request timeout (#284)
 - Connect() panics on concurrent schema update (#278)
+- Wrong Ttr setup by Queue.Cfg() (#278)
 
 ## [1.10.0] - 2022-12-31
 

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -29,6 +29,7 @@ Versioning](http://semver.org/spec/v2.0.0.html) except to the first release.
 - Connect() panics on concurrent schema update (#278)
 - Wrong Ttr setup by Queue.Cfg() (#278)
 - Flaky queue/Example_connectionPool (#278)
+- Flaky queue/Example_simpleQueueCustomMsgPack (#277)
 
 ## [1.10.0] - 2022-12-31
 

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -28,6 +28,7 @@ Versioning](http://semver.org/spec/v2.0.0.html) except to the first release.
 - Watcher events loss with a small per-request timeout (#284)
 - Connect() panics on concurrent schema update (#278)
 - Wrong Ttr setup by Queue.Cfg() (#278)
+- Flaky queue/Example_connectionPool (#278)
 
 ## [1.10.0] - 2022-12-31
 

--- a/Makefile
+++ b/Makefile
@@ -27,7 +27,7 @@ clean:
 .PHONY: deps
 deps: clean
 	( cd ./queue/testdata; $(TTCTL) rocks install queue 1.3.0 )
-	( cd ./crud/testdata; $(TTCTL) rocks install crud 1.0.0 )
+	( cd ./crud/testdata; $(TTCTL) rocks install crud 1.1.1 )
 
 .PHONY: datetime-timezones
 datetime-timezones:

--- a/Makefile
+++ b/Makefile
@@ -26,7 +26,7 @@ clean:
 
 .PHONY: deps
 deps: clean
-	( cd ./queue/testdata; $(TTCTL) rocks install queue 1.2.1 )
+	( cd ./queue/testdata; $(TTCTL) rocks install queue 1.3.0 )
 	( cd ./crud/testdata; $(TTCTL) rocks install crud 1.0.0 )
 
 .PHONY: datetime-timezones

--- a/config.lua
+++ b/config.lua
@@ -181,6 +181,23 @@ local function push_func(cnt)
 end
 rawset(_G, 'push_func', push_func)
 
+local function create_spaces()
+    for i=1,10 do
+        local s = box.schema.space.create('test' .. tostring(i), {
+            id = 700 + i,
+            if_not_exists = true,
+        })
+        local idx = s:create_index('test' .. tostring(i) .. 'primary', {
+            type = 'tree',
+            parts = {1, 'uint'},
+            if_not_exists = true
+        })
+        idx:drop()
+        s:drop()
+    end
+end
+rawset(_G, 'create_spaces', create_spaces)
+
 local function tarantool_version_at_least(wanted_major, wanted_minor, wanted_patch)
     -- https://github.com/tarantool/crud/blob/733528be02c1ffa3dacc12c034ee58c9903127fc/test/helper.lua#L316-L337
     local major_minor_patch = _TARANTOOL:split('-', 1)[1]

--- a/connection_pool/connection_pool_test.go
+++ b/connection_pool/connection_pool_test.go
@@ -33,7 +33,7 @@ var servers = []string{
 }
 
 var connOpts = tarantool.Opts{
-	Timeout: 500 * time.Millisecond,
+	Timeout: 5 * time.Second,
 	User:    "test",
 	Pass:    "test",
 }
@@ -2426,7 +2426,7 @@ func TestWatcher_Unregister_concurrent(t *testing.T) {
 func runTestMain(m *testing.M) int {
 	initScript := "config.lua"
 	waitStart := 100 * time.Millisecond
-	connectRetry := 3
+	connectRetry := 10
 	retryTimeout := 500 * time.Millisecond
 
 	// Tarantool supports streams and interactive transactions since version 2.10.0

--- a/crud/example_test.go
+++ b/crud/example_test.go
@@ -15,7 +15,7 @@ const (
 )
 
 var exampleOpts = tarantool.Opts{
-	Timeout: 500 * time.Millisecond,
+	Timeout: 5 * time.Second,
 	User:    "test",
 	Pass:    "test",
 }

--- a/crud/tarantool_test.go
+++ b/crud/tarantool_test.go
@@ -104,6 +104,33 @@ var object = crud.MapObject{
 	"name": "bla",
 }
 
+func connect(t testing.TB) *tarantool.Connection {
+	for i := 0; i < 10; i++ {
+		conn, err := tarantool.Connect(server, opts)
+		if err != nil {
+			t.Fatalf("Failed to connect: %s", err)
+		}
+
+		ret := struct {
+			_msgpack struct{} `msgpack:",asArray"` //nolint: structcheck,unused
+			Result   bool
+		}{}
+		err = conn.Do(tarantool.NewCall17Request("is_ready")).GetTyped(&ret)
+		if err != nil {
+			t.Fatalf("Failed to check is_ready: %s", err)
+		}
+
+		if ret.Result {
+			return conn
+		}
+
+		time.Sleep(time.Second)
+	}
+
+	t.Fatalf("Failed to wait for a ready state connect.")
+	return nil
+}
+
 var testProcessDataCases = []struct {
 	name            string
 	expectedRespLen int
@@ -454,7 +481,7 @@ func testCrudRequestCheck(t *testing.T, req tarantool.Request,
 }
 
 func TestCrudGenerateData(t *testing.T) {
-	conn := test_helpers.ConnectWithValidation(t, server, opts)
+	conn := connect(t)
 	defer conn.Close()
 
 	for _, testCase := range testGenerateDataCases {
@@ -477,7 +504,7 @@ func TestCrudGenerateData(t *testing.T) {
 }
 
 func TestCrudProcessData(t *testing.T) {
-	conn := test_helpers.ConnectWithValidation(t, server, opts)
+	conn := connect(t)
 	defer conn.Close()
 
 	for _, testCase := range testProcessDataCases {
@@ -527,7 +554,7 @@ func TestUnflattenRows(t *testing.T) {
 		tpls       []interface{}
 	)
 
-	conn := test_helpers.ConnectWithValidation(t, server, opts)
+	conn := connect(t)
 	defer conn.Close()
 
 	// Do `replace`.
@@ -586,7 +613,7 @@ func TestUnflattenRows(t *testing.T) {
 }
 
 func TestResultWithErr(t *testing.T) {
-	conn := test_helpers.ConnectWithValidation(t, server, opts)
+	conn := connect(t)
 	defer conn.Close()
 
 	for _, testCase := range testResultWithErrCases {
@@ -601,7 +628,7 @@ func TestResultWithErr(t *testing.T) {
 }
 
 func TestBoolResult(t *testing.T) {
-	conn := test_helpers.ConnectWithValidation(t, server, opts)
+	conn := connect(t)
 	defer conn.Close()
 
 	req := crud.MakeTruncateRequest(spaceName).Opts(baseOpts)
@@ -624,7 +651,7 @@ func TestBoolResult(t *testing.T) {
 }
 
 func TestNumberResult(t *testing.T) {
-	conn := test_helpers.ConnectWithValidation(t, server, opts)
+	conn := connect(t)
 	defer conn.Close()
 
 	req := crud.MakeCountRequest(spaceName).Opts(countOpts)
@@ -665,7 +692,7 @@ func TestBaseResult(t *testing.T) {
 		},
 	}
 
-	conn := test_helpers.ConnectWithValidation(t, server, opts)
+	conn := connect(t)
 	defer conn.Close()
 
 	req := crud.MakeSelectRequest(spaceName).Opts(selectOpts)
@@ -708,7 +735,7 @@ func TestManyResult(t *testing.T) {
 		},
 	}
 
-	conn := test_helpers.ConnectWithValidation(t, server, opts)
+	conn := connect(t)
 	defer conn.Close()
 
 	req := crud.MakeReplaceManyRequest(spaceName).Tuples(tuples).Opts(opManyOpts)
@@ -733,7 +760,7 @@ func TestManyResult(t *testing.T) {
 }
 
 func TestStorageInfoResult(t *testing.T) {
-	conn := test_helpers.ConnectWithValidation(t, server, opts)
+	conn := connect(t)
 	defer conn.Close()
 
 	req := crud.MakeStorageInfoRequest().Opts(baseOpts)

--- a/crud/tarantool_test.go
+++ b/crud/tarantool_test.go
@@ -20,7 +20,7 @@ var invalidSpaceName = "invalid"
 var indexNo = uint32(0)
 var indexName = "primary_index"
 var opts = tarantool.Opts{
-	Timeout: 500 * time.Millisecond,
+	Timeout: 5 * time.Second,
 	User:    "test",
 	Pass:    "test",
 }
@@ -31,7 +31,7 @@ var startOpts test_helpers.StartOpts = test_helpers.StartOpts{
 	User:         opts.User,
 	Pass:         opts.Pass,
 	WaitStart:    100 * time.Millisecond,
-	ConnectRetry: 3,
+	ConnectRetry: 10,
 	RetryTimeout: 500 * time.Millisecond,
 }
 

--- a/crud/testdata/config.lua
+++ b/crud/testdata/config.lua
@@ -59,6 +59,16 @@ s:create_index('bucket_id', {
     unique = false,
 })
 
+local function is_ready_false()
+	return false
+end
+
+local function is_ready_true()
+	return true
+end
+
+rawset(_G, 'is_ready', is_ready_false)
+
 -- Setup vshard.
 _G.vshard = vshard
 box.once('guest', function()
@@ -93,7 +103,5 @@ box.schema.user.grant('test', 'execute', 'universe', nil, { if_not_exists = true
 box.schema.user.grant('test', 'create,read,write,drop,alter', 'space', nil, { if_not_exists = true })
 box.schema.user.grant('test', 'create', 'sequence', nil, { if_not_exists = true })
 
--- Set listen only when every other thing is configured.
-box.cfg{
-    listen = os.getenv("TEST_TNT_LISTEN"),
-}
+-- Set is_ready = is_ready_true only when every other thing is configured.
+rawset(_G, 'is_ready', is_ready_true)

--- a/datetime/datetime_test.go
+++ b/datetime/datetime_test.go
@@ -37,7 +37,7 @@ var isDatetimeSupported = false
 
 var server = "127.0.0.1:3013"
 var opts = Opts{
-	Timeout: 500 * time.Millisecond,
+	Timeout: 5 * time.Second,
 	User:    "test",
 	Pass:    "test",
 }
@@ -1144,7 +1144,7 @@ func runTestMain(m *testing.M) int {
 		User:         opts.User,
 		Pass:         opts.Pass,
 		WaitStart:    100 * time.Millisecond,
-		ConnectRetry: 3,
+		ConnectRetry: 10,
 		RetryTimeout: 500 * time.Millisecond,
 	})
 	defer test_helpers.StopTarantoolWithCleanup(instance)

--- a/decimal/decimal_test.go
+++ b/decimal/decimal_test.go
@@ -616,7 +616,7 @@ func runTestMain(m *testing.M) int {
 		User:         opts.User,
 		Pass:         opts.Pass,
 		WaitStart:    100 * time.Millisecond,
-		ConnectRetry: 3,
+		ConnectRetry: 10,
 		RetryTimeout: 500 * time.Millisecond,
 	})
 	defer test_helpers.StopTarantoolWithCleanup(instance)

--- a/decimal/example_test.go
+++ b/decimal/example_test.go
@@ -22,7 +22,7 @@ import (
 func Example() {
 	server := "127.0.0.1:3013"
 	opts := tarantool.Opts{
-		Timeout:       500 * time.Millisecond,
+		Timeout:       5 * time.Second,
 		Reconnect:     1 * time.Second,
 		MaxReconnects: 3,
 		User:          "test",

--- a/example_test.go
+++ b/example_test.go
@@ -799,7 +799,7 @@ func ExampleConnection_Eval() {
 
 func ExampleConnect() {
 	conn, err := tarantool.Connect("127.0.0.1:3013", tarantool.Opts{
-		Timeout:     500 * time.Millisecond,
+		Timeout:     5 * time.Second,
 		User:        "test",
 		Pass:        "test",
 		Concurrency: 32,
@@ -895,11 +895,9 @@ func ExampleConnection_Execute() {
 	}
 	server := "127.0.0.1:3013"
 	opts := tarantool.Opts{
-		Timeout:       500 * time.Millisecond,
-		Reconnect:     1 * time.Second,
-		MaxReconnects: 3,
-		User:          "test",
-		Pass:          "test",
+		Timeout: 5 * time.Second,
+		User:    "test",
+		Pass:    "test",
 	}
 	client, err := tarantool.Connect(server, opts)
 	if err != nil {
@@ -1015,11 +1013,9 @@ func ExampleConnection_NewPrepared() {
 
 	server := "127.0.0.1:3013"
 	opts := tarantool.Opts{
-		Timeout:       500 * time.Millisecond,
-		Reconnect:     1 * time.Second,
-		MaxReconnects: 3,
-		User:          "test",
-		Pass:          "test",
+		Timeout: 5 * time.Second,
+		User:    "test",
+		Pass:    "test",
 	}
 	conn, err := tarantool.Connect(server, opts)
 	if err != nil {
@@ -1057,8 +1053,8 @@ func ExampleConnection_NewWatcher() {
 
 	server := "127.0.0.1:3013"
 	opts := tarantool.Opts{
-		Timeout:       500 * time.Millisecond,
-		Reconnect:     1 * time.Second,
+		Timeout:       5 * time.Second,
+		Reconnect:     5 * time.Second,
 		MaxReconnects: 3,
 		User:          "test",
 		Pass:          "test",

--- a/multi/example_test.go
+++ b/multi/example_test.go
@@ -9,7 +9,7 @@ import (
 
 func ExampleConnect() {
 	multiConn, err := Connect([]string{"127.0.0.1:3031", "127.0.0.1:3032"}, tarantool.Opts{
-		Timeout: 500 * time.Millisecond,
+		Timeout: 5 * time.Second,
 		User:    "test",
 		Pass:    "test",
 	})
@@ -21,7 +21,7 @@ func ExampleConnect() {
 
 func ExampleConnectWithOpts() {
 	multiConn, err := ConnectWithOpts([]string{"127.0.0.1:3301", "127.0.0.1:3302"}, tarantool.Opts{
-		Timeout: 500 * time.Millisecond,
+		Timeout: 5 * time.Second,
 		User:    "test",
 		Pass:    "test",
 	}, OptsMulti{

--- a/multi/multi_test.go
+++ b/multi/multi_test.go
@@ -20,7 +20,7 @@ var spaceNo = uint32(617)
 var spaceName = "test"
 var indexNo = uint32(0)
 var connOpts = tarantool.Opts{
-	Timeout: 500 * time.Millisecond,
+	Timeout: 5 * time.Second,
 	User:    "test",
 	Pass:    "test",
 }
@@ -611,7 +611,7 @@ func TestConnectionMulti_NewWatcher(t *testing.T) {
 func runTestMain(m *testing.M) int {
 	initScript := "config.lua"
 	waitStart := 100 * time.Millisecond
-	connectRetry := 3
+	connectRetry := 10
 	retryTimeout := 500 * time.Millisecond
 
 	// Tarantool supports streams and interactive transactions since version 2.10.0

--- a/queue/example_connection_pool_test.go
+++ b/queue/example_connection_pool_test.go
@@ -156,12 +156,12 @@ func Example_connectionPool() {
 		"127.0.0.1:3015",
 	}
 	connOpts := tarantool.Opts{
-		Timeout: 1 * time.Second,
+		Timeout: 5 * time.Second,
 		User:    "test",
 		Pass:    "test",
 	}
 	poolOpts := connection_pool.OptsPool{
-		CheckTimeout:      1 * time.Second,
+		CheckTimeout:      5 * time.Second,
 		ConnectionHandler: h,
 	}
 	connPool, err := connection_pool.ConnectWithOpts(servers, connOpts, poolOpts)

--- a/queue/example_connection_pool_test.go
+++ b/queue/example_connection_pool_test.go
@@ -61,6 +61,13 @@ func (h *QueueConnectionHandler) Discovered(conn *tarantool.Connection,
 
 	// Set up a queue module configuration for an instance.
 	q := queue.New(conn, h.name)
+
+	// Set up a queue module configuration for an instance. Ideally, this
+	// should be done before box.cfg({}) or you need to wait some time
+	// before start a work.
+	//
+	// See:
+	// https://github.com/tarantool/queue/issues/206
 	opts := queue.CfgOpts{InReplicaset: true, Ttr: 60 * time.Second}
 
 	if h.err = q.Cfg(opts); h.err != nil {

--- a/queue/example_connection_pool_test.go
+++ b/queue/example_connection_pool_test.go
@@ -18,12 +18,12 @@ type QueueConnectionHandler struct {
 	name string
 	cfg  queue.Cfg
 
-	uuid          uuid.UUID
-	registered    bool
-	err           error
-	mutex         sync.Mutex
-	masterUpdated chan struct{}
-	masterCnt     int32
+	uuid       uuid.UUID
+	registered bool
+	err        error
+	mutex      sync.Mutex
+	updated    chan struct{}
+	masterCnt  int32
 }
 
 // QueueConnectionHandler implements the ConnectionHandler interface.
@@ -32,9 +32,9 @@ var _ connection_pool.ConnectionHandler = &QueueConnectionHandler{}
 // NewQueueConnectionHandler creates a QueueConnectionHandler object.
 func NewQueueConnectionHandler(name string, cfg queue.Cfg) *QueueConnectionHandler {
 	return &QueueConnectionHandler{
-		name:          name,
-		cfg:           cfg,
-		masterUpdated: make(chan struct{}, 10),
+		name:    name,
+		cfg:     cfg,
+		updated: make(chan struct{}, 10),
 	}
 }
 
@@ -53,14 +53,23 @@ func (h *QueueConnectionHandler) Discovered(conn *tarantool.Connection,
 	}
 
 	master := role == connection_pool.MasterRole
-	if master {
-		defer func() {
-			h.masterUpdated <- struct{}{}
-		}()
+
+	q := queue.New(conn, h.name)
+
+	// Check is queue ready to work.
+	if state, err := q.State(); err != nil {
+		h.updated <- struct{}{}
+		h.err = err
+		return err
+	} else if master && state != queue.RunningState {
+		return fmt.Errorf("queue state is not RUNNING: %d", state)
+	} else if !master && state != queue.InitState && state != queue.WaitingState {
+		return fmt.Errorf("queue state is not INIT and not WAITING: %d", state)
 	}
 
-	// Set up a queue module configuration for an instance.
-	q := queue.New(conn, h.name)
+	defer func() {
+		h.updated <- struct{}{}
+	}()
 
 	// Set up a queue module configuration for an instance. Ideally, this
 	// should be done before box.cfg({}) or you need to wait some time
@@ -79,10 +88,6 @@ func (h *QueueConnectionHandler) Discovered(conn *tarantool.Connection,
 		return nil
 	}
 
-	if h.err = q.Create(h.cfg); h.err != nil {
-		return h.err
-	}
-
 	if !h.registered {
 		// We register a shared session at the first time.
 		if h.uuid, h.err = q.Identify(nil); h.err != nil {
@@ -94,6 +99,10 @@ func (h *QueueConnectionHandler) Discovered(conn *tarantool.Connection,
 		if _, h.err = q.Identify(&h.uuid); h.err != nil {
 			return h.err
 		}
+	}
+
+	if h.err = q.Create(h.cfg); h.err != nil {
+		return h.err
 	}
 
 	fmt.Printf("Master %s is ready to work!\n", conn.Addr())
@@ -113,7 +122,7 @@ func (h *QueueConnectionHandler) Deactivated(conn *tarantool.Connection,
 
 // Closes closes a QueueConnectionHandler object.
 func (h *QueueConnectionHandler) Close() {
-	close(h.masterUpdated)
+	close(h.updated)
 }
 
 // Example demonstrates how to use the queue package with the connection_pool
@@ -162,8 +171,10 @@ func Example_connectionPool() {
 	}
 	defer connPool.Close()
 
-	// Wait for a master instance identification in the queue.
-	<-h.masterUpdated
+	// Wait for a queue initialization and master instance identification in
+	// the queue.
+	<-h.updated
+	<-h.updated
 	if h.err != nil {
 		fmt.Printf("Unable to identify in the pool: %s", h.err)
 		return
@@ -184,14 +195,17 @@ func Example_connectionPool() {
 
 	// Switch a master instance in the pool.
 	roles := []bool{true, false}
-	err = test_helpers.SetClusterRO(servers, connOpts, roles)
-	if err != nil {
-		fmt.Printf("Unable to set cluster roles: %s", err)
-		return
+	for {
+		err := test_helpers.SetClusterRO(servers, connOpts, roles)
+		if err == nil {
+			break
+		}
 	}
 
-	// Wait for a new master instance re-identification.
-	<-h.masterUpdated
+	// Wait for a replica instance connection and a new master instance
+	// re-identification.
+	<-h.updated
+	<-h.updated
 	h.mutex.Lock()
 	err = h.err
 	h.mutex.Unlock()
@@ -211,17 +225,24 @@ func Example_connectionPool() {
 		time.Sleep(poolOpts.CheckTimeout)
 	}
 
-	// Take a data from the new master instance.
-	task, err := q.Take()
-	if err != nil {
-		fmt.Println("Unable to got task:", err)
-	} else if task == nil {
-		fmt.Println("task == nil")
-	} else if task.Data() == nil {
-		fmt.Println("task.Data() == nil")
-	} else {
-		task.Ack()
-		fmt.Println("Got data:", task.Data())
+	for {
+		// Take a data from the new master instance.
+		task, err := q.Take()
+
+		if err == connection_pool.ErrNoRwInstance {
+			// It may be not registered yet by the pool.
+			continue
+		} else if err != nil {
+			fmt.Println("Unable to got task:", err)
+		} else if task == nil {
+			fmt.Println("task == nil")
+		} else if task.Data() == nil {
+			fmt.Println("task.Data() == nil")
+		} else {
+			task.Ack()
+			fmt.Println("Got data:", task.Data())
+		}
+		break
 	}
 
 	// Output:

--- a/queue/example_msgpack_test.go
+++ b/queue/example_msgpack_test.go
@@ -48,7 +48,7 @@ func (c *dummyData) EncodeMsgpack(e *encoder) error {
 func Example_simpleQueueCustomMsgPack() {
 	opts := tarantool.Opts{
 		Reconnect:     time.Second,
-		Timeout:       2500 * time.Millisecond,
+		Timeout:       5 * time.Second,
 		MaxReconnects: 5,
 		User:          "test",
 		Pass:          "test",
@@ -65,9 +65,9 @@ func Example_simpleQueueCustomMsgPack() {
 		IfNotExists: true,
 		Kind:        queue.FIFO,
 		Opts: queue.Opts{
-			Ttl:   10 * time.Second,
-			Ttr:   5 * time.Second,
-			Delay: 3 * time.Second,
+			Ttl:   20 * time.Second,
+			Ttr:   10 * time.Second,
+			Delay: 6 * time.Second,
 			Pri:   1,
 		},
 	}

--- a/queue/queue.go
+++ b/queue/queue.go
@@ -143,7 +143,7 @@ func (opts CfgOpts) toMap() map[string]interface{} {
 	ret := make(map[string]interface{})
 	ret["in_replicaset"] = opts.InReplicaset
 	if opts.Ttr != 0 {
-		ret["ttr"] = opts.Ttr
+		ret["ttr"] = opts.Ttr.Seconds()
 	}
 	return ret
 }

--- a/queue/queue_test.go
+++ b/queue/queue_test.go
@@ -22,7 +22,7 @@ var serversPool = []string{
 var instances []test_helpers.TarantoolInstance
 
 var opts = Opts{
-	Timeout: 2500 * time.Millisecond,
+	Timeout: 5 * time.Second,
 	User:    "test",
 	Pass:    "test",
 	//Concurrency: 32,
@@ -913,7 +913,7 @@ func runTestMain(m *testing.M) int {
 		User:         opts.User,
 		Pass:         opts.Pass,
 		WaitStart:    100 * time.Millisecond,
-		ConnectRetry: 3,
+		ConnectRetry: 10,
 		RetryTimeout: 500 * time.Millisecond,
 	})
 

--- a/queue/queue_test.go
+++ b/queue/queue_test.go
@@ -938,13 +938,22 @@ func runTestMain(m *testing.M) int {
 
 	defer test_helpers.StopTarantoolInstances(instances)
 
-	roles := []bool{false, true}
-	connOpts := Opts{
-		Timeout: 500 * time.Millisecond,
-		User:    "test",
-		Pass:    "test",
+	for i := 0; i < 10; i++ {
+		// We need to skip bootstrap errors and to make sure that cluster is
+		// configured.
+		roles := []bool{false, true}
+		connOpts := Opts{
+			Timeout: 500 * time.Millisecond,
+			User:    "test",
+			Pass:    "test",
+		}
+
+		err = test_helpers.SetClusterRO(serversPool, connOpts, roles)
+		if err == nil {
+			break
+		}
+		time.Sleep(time.Second)
 	}
-	err = test_helpers.SetClusterRO(serversPool, connOpts, roles)
 
 	if err != nil {
 		log.Fatalf("Failed to set roles in tarantool pool: %s", err)

--- a/queue/testdata/pool.lua
+++ b/queue/testdata/pool.lua
@@ -25,6 +25,9 @@ end
 
 local queue = require('queue')
 rawset(_G, 'queue', queue)
+-- queue.cfg({in_replicaset = true}) should be called before box.cfg({})
+-- https://github.com/tarantool/queue/issues/206
+queue.cfg({in_replicaset = true, ttr = 60})
 
 local listen = os.getenv("TEST_TNT_LISTEN")
 box.cfg{

--- a/schema.go
+++ b/schema.go
@@ -325,8 +325,13 @@ func (conn *Connection) loadSchema() (err error) {
 		return err
 	}
 	for _, index := range indexes {
-		schema.SpacesById[index.SpaceId].IndexesById[index.Id] = index
-		schema.SpacesById[index.SpaceId].Indexes[index.Name] = index
+		spaceId := index.SpaceId
+		if _, ok := schema.SpacesById[spaceId]; ok {
+			schema.SpacesById[spaceId].IndexesById[index.Id] = index
+			schema.SpacesById[spaceId].Indexes[index.Name] = index
+		} else {
+			return errors.New("concurrent schema update")
+		}
 	}
 
 	conn.lockShards()

--- a/settings/tarantool_test.go
+++ b/settings/tarantool_test.go
@@ -19,7 +19,7 @@ var isSettingsSupported = false
 
 var server = "127.0.0.1:3013"
 var opts = tarantool.Opts{
-	Timeout: 500 * time.Millisecond,
+	Timeout: 5 * time.Second,
 	User:    "test",
 	Pass:    "test",
 }
@@ -637,7 +637,7 @@ func runTestMain(m *testing.M) int {
 		User:         opts.User,
 		Pass:         opts.Pass,
 		WaitStart:    100 * time.Millisecond,
-		ConnectRetry: 3,
+		ConnectRetry: 10,
 		RetryTimeout: 500 * time.Millisecond,
 	})
 

--- a/shutdown_test.go
+++ b/shutdown_test.go
@@ -418,9 +418,12 @@ func TestGracefulShutdownCloseConcurrent(t *testing.T) {
 
 			// Do not wait till Tarantool register out watcher,
 			// test everything is ok even on async.
-
-			conn := test_helpers.ConnectWithValidation(t, shtdnServer, shtdnClntOpts)
-			defer conn.Close()
+			conn, err := Connect(shtdnServer, shtdnClntOpts)
+			if err != nil {
+				t.Errorf("Failed to connect: %s", err)
+			} else {
+				defer conn.Close()
+			}
 
 			// Wait till all connections created.
 			srvToStop.Done()

--- a/shutdown_test.go
+++ b/shutdown_test.go
@@ -22,7 +22,7 @@ var shtdnClntOpts = Opts{
 	User:                 opts.User,
 	Pass:                 opts.Pass,
 	Timeout:              20 * time.Second,
-	Reconnect:            200 * time.Millisecond,
+	Reconnect:            500 * time.Millisecond,
 	MaxReconnects:        10,
 	RequiredProtocolInfo: ProtocolInfo{Features: []ProtocolFeature{WatchersFeature}},
 }
@@ -32,7 +32,7 @@ var shtdnSrvOpts = test_helpers.StartOpts{
 	User:         shtdnClntOpts.User,
 	Pass:         shtdnClntOpts.Pass,
 	WaitStart:    100 * time.Millisecond,
-	ConnectRetry: 3,
+	ConnectRetry: 10,
 	RetryTimeout: 500 * time.Millisecond,
 }
 

--- a/ssl_test.go
+++ b/ssl_test.go
@@ -130,7 +130,7 @@ func serverTnt(serverOpts, clientOpts SslOpts, auth Auth) (test_helpers.Tarantoo
 		User:            "test",
 		Pass:            "test",
 		WaitStart:       100 * time.Millisecond,
-		ConnectRetry:    3,
+		ConnectRetry:    10,
 		RetryTimeout:    500 * time.Millisecond,
 	})
 }

--- a/tarantool_test.go
+++ b/tarantool_test.go
@@ -3894,6 +3894,29 @@ func TestWatcher_Unregister_concurrent(t *testing.T) {
 	wg.Wait()
 }
 
+func TestConnect_schema_update(t *testing.T) {
+	conn := test_helpers.ConnectWithValidation(t, server, opts)
+	defer conn.Close()
+
+	for i := 0; i < 100; i++ {
+		fut := conn.Do(NewCallRequest("create_spaces"))
+
+		if conn, err := Connect(server, opts); err != nil {
+			if err.Error() != "concurrent schema update" {
+				t.Errorf("unexpected error: %s", err)
+			}
+		} else if conn == nil {
+			t.Errorf("conn is nil")
+		} else {
+			conn.Close()
+		}
+
+		if _, err := fut.Get(); err != nil {
+			t.Errorf("Failed to call create_spaces: %s", err)
+		}
+	}
+}
+
 // runTestMain is a body of TestMain function
 // (see https://pkg.go.dev/testing#hdr-Main).
 // Using defer + os.Exit is not works so TestMain body

--- a/tarantool_test.go
+++ b/tarantool_test.go
@@ -28,7 +28,7 @@ var startOpts test_helpers.StartOpts = test_helpers.StartOpts{
 	User:         opts.User,
 	Pass:         opts.Pass,
 	WaitStart:    100 * time.Millisecond,
-	ConnectRetry: 3,
+	ConnectRetry: 10,
 	RetryTimeout: 500 * time.Millisecond,
 }
 
@@ -75,7 +75,7 @@ var spaceName = "test"
 var indexNo = uint32(0)
 var indexName = "primary"
 var opts = Opts{
-	Timeout: 500 * time.Millisecond,
+	Timeout: 5 * time.Second,
 	User:    "test",
 	Pass:    "test",
 	//Concurrency: 32,
@@ -3580,7 +3580,7 @@ func TestConnection_NewWatcher_reconnect(t *testing.T) {
 		User:         opts.User,
 		Pass:         opts.Pass,
 		WaitStart:    100 * time.Millisecond,
-		ConnectRetry: 3,
+		ConnectRetry: 10,
 		RetryTimeout: 500 * time.Millisecond,
 	})
 	defer test_helpers.StopTarantoolWithCleanup(inst)

--- a/uuid/uuid_test.go
+++ b/uuid/uuid_test.go
@@ -20,7 +20,7 @@ var isUUIDSupported = false
 
 var server = "127.0.0.1:3013"
 var opts = Opts{
-	Timeout: 500 * time.Millisecond,
+	Timeout: 5 * time.Second,
 	User:    "test",
 	Pass:    "test",
 }
@@ -158,7 +158,7 @@ func runTestMain(m *testing.M) int {
 		User:         opts.User,
 		Pass:         opts.Pass,
 		WaitStart:    100 * time.Millisecond,
-		ConnectRetry: 3,
+		ConnectRetry: 10,
 		RetryTimeout: 500 * time.Millisecond,
 	})
 	defer test_helpers.StopTarantoolWithCleanup(inst)

